src/PJ_helmert.c

```diff
@@ -101,7 +110,206 @@ struct pj_opaque_helmert {
 #define R21 (Q->R[2][1])
 #define R22 (Q->R[2][2])
 
+static void update_parameters(PJ *P) {
+    /**************************************************************************
+
+    Update transformation parameters.
+    ---------------------------------
+
+    The 14-parameter Helmert transformation is at it's core the same as the
+    7-parameter transformation, since the transformation parameters are
+    projected forward or backwards in time via the rate of changes of the
+    parameters. The transformation parameters are calculated for a specific
+    epoch before the actual Helmert transformation is carried out.
+
+    The transformation parameters are updated with the following equation [0]:
+
+                      .
+    P(t) = P(EPOCH) + P * (t - EPOCH)
+
+                                                              .
+    where EPOCH is the epoch indicated in the above table and P is the rate
+    of that parameter.
+
+    [0] http://itrf.ign.fr/doc_ITRF/Transfo-ITRF2008_ITRFs.txt
+    **************************************************************************/
+
+    struct pj_opaque_helmert *Q = (struct pj_opaque_helmert *) P->opaque;
+    double dt = Q->t_obs - Q->epoch;
+
+    Q->xyz.x = Q->xyz_0.x + Q->dxyz.x * dt;
+    Q->xyz.y = Q->xyz_0.y + Q->dxyz.y * dt;
+    Q->xyz.z = Q->xyz_0.z + Q->dxyz.z * dt;
+
+    Q->opk.o = Q->opk_0.o + Q->dopk.o * dt;
+    Q->opk.p = Q->opk_0.p + Q->dopk.p * dt;
+    Q->opk.k = Q->opk_0.k + Q->dopk.k * dt;
+
+    Q->scale = Q->scale_0 + Q->dscale * dt;
+
+    /* debugging output */
+    pj_log_trace(P, "Transformation parameters for observation epoch %g:", Q->t_obs);
```

Comment:
Since this is a large block of logging calls, I suggest surrounding them with
"if (pj_err_level (P, PJ_ERR_TELL) >= PJ_LOG_TRACE) {...}". It will eliminate 8 function calls (although, since this is setup-code, it will happen only once. The situation is worse if it's part of the projection code)